### DISCUSSION TITLE
why not shared registry between main and shared library???

### DISCUSSION THREAD

**Discussion by hbwang15** [gabime/spdlog#2033] [Q&A] [answered]
> Because spdlog is header-only, building shared libraries and using it in a main program will not share the registry between them. This means that calls to functions like spdlog::set_level(spdlog::level::level_enum::info) will not change loggers in DLLs.

why not shared registry between main and shared library???

**Comment by tt4g** [comment, marked as answer]
This can be described as a C/C++ language specification and security limitation.

Since a shared library may be linked and used by multiple applications, the OS reserves memory storage dedicated to the library when the library is loaded into memory.
The objects generated by the library code are placed in the memory storage reserved for the library (this allows objects generated by a single library to be shared by multiple applications).

The library's memory storage cannot be accessed directly by different applications (this is a memory access violation).

See also:
  * https://stackoverflow.com/questions/1178127/shared-libraries-memory-space
  * https://unix.stackexchange.com/questions/116327/loading-of-shared-libraries-and-ram-usage

**Reply by tt4g** [reply]
> why not shared registry between main and shared library???

it may be because it is very difficult to share objects with complex structures such as the `extern` variable.

The spdlog registry contains implementations of `spdog::logger` and `spdlog::sinks::sink` (and their subclasses).
When `delete` these objects, each class destructor has to be called.

The shared library does not know how to `delete` a subclass that the application has implemented on its own. There should be an instruction in the application code that corresponds to the destructor, but calling this code from the shared library should be an access violation.

**Reply by hbwang15** [reply]
thank you for your reply. i still be confused of static local member in shared lib and main , as you said it not shared memory, but when i build the dynamic lib from spdlog example. i got an exception from spdlog which shouldn't occur when register not shared.
please help me to solve this problem. The recipe code in github link is : [https://github.com/hbwang15/spdlogInSharedLib](url)

you can find the process how i build shared lib and use it in main from Readme.   and complier and system info is : 

> g++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE

> 18.04.1-Ubuntu SMP Tue Jul 13 19:40:02 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

**Reply by tt4g** [reply]
The same logger has been registered twice with `spdlog::register_logger()`.

* https://github.com/hbwang15/spdlogInSharedLib/blob/65bdec50d0d3a926dbe6e6f5f6443519543a3922/lib.cpp#L23
* https://github.com/hbwang15/spdlogInSharedLib/blob/65bdec50d0d3a926dbe6e6f5f6443519543a3922/main.cpp#L20

Loggers in the registry are identified by `spdlog::logger::name()`, so loggers with the same name cannot be registered.

https://github.com/gabime/spdlog/blob/v1.9.1/include/spdlog/details/registry-inl.h#L301

You need to remove either `spdlog::register_logger()`.

**Reply by hbwang15** [reply]
but the example is from [https://github.com/gabime/spdlog/wiki/How-to-use-spdlog-in-DLLs](url)

**Reply by hbwang15** [reply]
i am so confused by your explain, if the dll is not shared the register with the main, why exception happens

**Reply by tt4g** [reply]
I'm sorry. It seems that my explanation was insufficient.

I think your first question was written after looking at the Wiki: https://github.com/gabime/spdlog/wiki/How-to-use-spdlog-in-DLLs
The content here contains notes on using spdlog inside DLLs on Windows OS.
In other words, it is not relevant for general UNIX Like OS.
Please consider that my answer so far is also mainly about memory space in Windows OS.

I assume that your code is built and running on a UNIX Like OS.
As far as I know, in that environment, the shared library (`.so`) and the application share the `static` variable.

For more information, please browse this Stack Overflow post: https://stackoverflow.com/questions/19373061/what-happens-to-global-and-static-variables-in-a-shared-library-when-it-is-dynam

**Reply by hbwang15** [reply]
Thank you very much for your answers
